Installation [8381] (#151)
eProsima/Fast-RTPS-docs: docs/installation/sources/sources_linux.rst

.. _linux_sources:

Linux installation from sources
===============================

To install *eProsima Fast DDS* from sources, you first need to meet the required dependencies

> (see requirements_linux_sources_)

Author: ok. Don't know why it's not working.

and then choose whether to follow either the colcon_ (see colcon colcon_installation_linux_) or the CMake_
(see cmake_installation_linux_) installation instructions.

.. _requirements_linux_sources:

Requirements
------------

*eProsima Fast DDS* requires the following dependencies when building from sources in a Linux environment.

Asio and TinyXML2 libraries
^^^^^^^^^^^^^^^^^^^^^^^^^^^

You can install these libraries using the package manager of your Linux distribution.
For example, on Ubuntu you can install them with the command:

.. code-block:: bash

    sudo apt install libasio-dev libtinyxml2-dev

Python
^^^^^^

> Download and install the Python_ version that better fits your requirements.

Author: for installing pip and vcstools.

CMake, g++, pip, wget and git
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

You can install CMake_, g++, pip_, wget_ and git_ using the package manager of your Linux distribution.
For example, on Ubuntu you can install them with the command:

.. code-block:: bash

    sudo apt install cmake g++ pip wget git

OpenSSL
^^^^^^^

You can install OpenSSL_ using the package manager of your Linux distribution.
For example, on Ubuntu you can install it with the command:

.. code-block:: bash

   sudo apt install libssl-dev

Gtest
^^^^^

> .. note::

Reviewer: I don't think it's clear enough that you only need Gtest if compiling tests

Author: I removed the note, now it's plain text.
what else?

    By default, *eProsima Fast DDS* does not compile tests.

>     You can activate them by adding the :code:`-DPERFORMANCE_TESTS=ON` flag when calling colcon_ or CMake_

Reviewer: There are many more tests than performance. I'd simply refer to the cmake options section instead of listing all the options that enable tests.

Author: ok.

    (for details, see below).

> You can find information on how to install Gtest at this `link <https://github.com/google/googletest>`_.

Reviewer: ```suggestion
For a detailed description of the installation process, please refer to `Gtest Installation Guide <https://github.com/google/googletest>_`.
```

Author: ok.

.. _colcon_installation_linux:

Colcon Installation
-------------------

colcon_ is a command line tool to build sets of software packages.

> This section explains how to use it to compile easily *eProsima Fast DDS* and its dependencies.

Reviewer: ```suggestion
This section explains how to use it to compile *eProsima Fast DDS* and its dependencies.
```

Author: ok.

> First, install the ROS2 development tools (colcon_ and vcstool_):

Reviewer: In general
```suggestion
First, install the ROS 2 development tools (colcon_ and vcstool_):
```

Author: ok.

> First, install the ROS2 development tools (colcon_ and vcstool_):

Reviewer: All the steps in an enumerated list

Author: Are you sure? All the rest of the installation doesn't have this philosophy!

Author: put numbers in colcon and cmake sections.
use the #. command

.. code-block:: bash

    pip install -U colcon-common-extensions vcstool

.. note::

>     If this fails due to an Environment Error, add the :code:`--user` flag to your installation.

Reviewer: I don't know hot to do this

Now, create a colcon_ workspace and download the repos file that will be used to install *eProsima Fast DDS* and
its dependencies:

.. code-block:: bash

>     $ mkdir Fast-DDS-ws && cd Fast-DDS-ws

Reviewer: Not very fond of the workspace name. If changed, propagate everywhere
```suggestion
    $ mkdir fastdds_ws && cd fastdds_ws
```

Author: ok.
I removed the 'workspace' keyword and changed it to 'Fast-DDS'.

>     $ wget https://raw.githubusercontent.com/eProsima/Fast-RTPS/master/fastrtps.repos

Reviewer: ```suggestion
    $ wget https://raw.githubusercontent.com/eProsima/Fast-DDS/master/fastrtps.repos
```

Author: ok.

    $ mkdir src
    $ vcs import src < fastrtps.repos

> Finally, use colcon_ to compile all software:

Reviewer: The user would like to know how to pass CMake options to `colcon build`. We should refer to that section of colcon's docs

Author: ok.

.. code-block:: bash

    colcon build

Once that’s finished building, you can source the new colcon overlay:

.. code-block:: bash

    source install/setup.bash

.. important::

    The sourcing of the local colcon overlay is required every time the colcon workspace is opened in a new shell
    environment to run an *eProsima Fast DDS* instance.
    As an alternative, you can add it permanently to you path by typing the following:

    .. code-block:: bash

        echo 'source PATH_TO_WORKSPACE/Fast-DDS-ws/install/setup.bash' >> ~/.bashrc

    Where :code:`PATH_TO_WORKSPACE` is the path to the :code:`Fast-DDS-ws` worskspace.

.. note::

    If you want to compile the examples, you will need to add the flag
    :code:`--cmake-args "-DCOMPILE_EXAMPLES=ON"` when running :code:`colcon build`.
    If you want to compile the performance tests, you will need to add the flag
    :code:`--cmake-args "--DPERFORMANCE_TESTS=ON"` when running :code:`colcon build`.

>     For this step, you need Gtest_ as explained in the requirements_linux_sources_ section above.

Reviewer: In general, this will be covered in the cmake options section, we don't need it here

Author: ok. Deleted.

.. _cmake_installation_linux:

CMake Installation
------------------

This section explains how to compile *eProsima Fast DDS* locally with CMake_.
First of all, create a Fast-DDS directory where to download and build *eProsima Fast DDS* and its dependencies:

.. code-block:: bash

>     mkdir Fast-DDS && cd Fast-DDS

Reviewer: In general, I'll avoid running two instructions in one command. Moreover, I'd specify the user dir for the installation.
```suggestion
    mkdir ~/Fast-DDS
    cd ~/Fast-DDS
```

Author: ok

Now clone the following dependencies and compile them using CMake_.

* `Foonathan memory <https://github.com/foonathan/memory>`_

  .. code-block:: bash

>       $ git clone https://github.com/eProsima/foonathan_memory_vendor.git

Reviewer: In general, I'd remove all the `$` from the code blocks. I know some people do it, but we are not doing it elsewhere (not even in this chapter) in the docs and it does not bring anything. Moreover, I'd always start by changing to the appropriate directory
```suggestion
       cd ~/Fast-DDS
      git clone https://github.com/eProsima/foonathan_memory_vendor.git
```

Author: It gave me an error, without the $.

Author: check again.

Author: taken the $ away.

      $ mkdir foonathan_memory_vendor/build && cd foonathan_memory_vendor/build

>       $ cmake .. -DCMAKE_INSTALL_PREFIX=../../install

Reviewer: Please use full paths

Author: ok.

      $ sudo cmake --build . --target install

>       $ cd ../..

Reviewer: In general
```suggestion
```

Author: ok

* `Fast CDR <https://github.com/eProsima/Fast-CDR.git>`_

  .. code-block:: bash

      $ git clone https://github.com/eProsima/Fast-CDR.git
      $ mkdir Fast-CDR/build && cd Fast-CDR/build
      $ cmake .. -DCMAKE_INSTALL_PREFIX=../../install
      $ sudo cmake --build . --target install
      $ cd ../..

Once all dependencies are installed, you will be able to compile and install *eProsima Fast DDS*:

.. code-block:: bash

    $ git clone https://github.com/eProsima/Fast-RTPS.git
    $ mkdir Fast-RTPS/build && cd Fast-RTPS/build
    $ cmake ..  -DCMAKE_INSTALL_PREFIX=../../install -DCMAKE_PREFIX_PATH=../../install
    $ sudo cmake --build . --target install

> 

Reviewer: If there are 2 different things, I think they belong to different subsections

Author: ??

Author: yes. divide local and system-wide -> do not repeat all steps.

If you want to install *eProsima Fast DDS* system-wide instead of locally, you need to remove all the flags that
appear in the configuration steps of :code:`Fast-CDR` and :code:`Fast-RTPS`, and change the one in the
configuration step of :code:`foonathan_memory_vendor` to the following:

.. code-block:: bash

>     -DCMAKE_INSTALL_PREFIX=/usr/local/ -DBUILD_SHARED_LIBS=ON

Reviewer: I'd actually put `-DBUILD_SHARED_LIBS=ON` in the local installation as well.

Author: only for foonathan?

Author: yes. adjust text of system-wide part accordingly.

Author: ok.

.. note::

    If you want to compile the examples, you will need to add the argument :code:`-DCOMPILE_EXAMPLES=ON` when calling
    the configuration CMake_.
    If you want to compile the performance tests, you will need to add the argument
    :code:`-DPERFORMANCE_TESTS=ON` when calling the configuration CMake_.
    For this step, you need Gtest_ as explained in the requirements_linux_sources_ section above.

.. important::

>     When running an *eProsima Fast DDS* application, you need to link it with the library :code:`/usr/local/lib/`

Reviewer: ```suggestion
    When using *eProsima Fast DDS* in an application, you need to link it with the library :code:`/usr/local/lib/`
```

Author: ok.

>     When running an *eProsima Fast DDS* application, you need to link it with the library :code:`/usr/local/lib/`

Reviewer: I think this sentence is weird; the library is not `/usr/local/lib/`, that is merely the directory where the library is.

Author: ok. changed from:
the library :code:`/usr/local/lib/` 
to:
the library in :code:`/usr/local/lib/`

    where the packages have been installed. You can either prepare the environment locally by typing the command:

    .. code-block:: bash

        export LD_LIBRARY_PATH=/usr/local/lib/

>     in the console you use to run the *eProsima Fast DDS* instance, or permanently add it to your path, by typing:

Reviewer: Some part of this sentence belong to the previous

Author: yes.

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

These packages provide the tools required to install *eProsima Fast DDS* and its dependencies from command line.
Install CMake_, `g++ <https://gcc.gnu.org/>`_, pip_, wget_ and git_ using the package manager of the appropriate
Linux distribution. For example, on Ubuntu use the command:

.. code-block:: bash

    sudo apt install cmake g++ pip wget git

.. _gtest_sl:

Gtest
^^^^^

GTest is a unit testing library for C++.

By default, *eProsima Fast DDS* does not compile tests.
It is possible to activate them with the opportune
`CMake configuration options <https://cmake.org/cmake/help/v3.6/manual/cmake.1.html#options>`_
when calling colcon_ or CMake_.
For more details, please refer to the :ref:`cmake_options` section.

For a detailed description of the Gtest installation process, please refer to the
`Gtest Installation Guide <https://github.com/google/googletest>`_.

.. _dependencies_sl:

Dependencies
------------

*eProsima Fast RTPS* has the following dependencies, when installed from binaries in a Linux environment:

* :ref:`asiotinyxml2_sl`
* :ref:`openssl_sl`

.. _asiotinyxml2_sl:

Asio and TinyXML2 libraries
^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Asio is a cross-platform C++ library for network and low-level I/O programming providing with a consistent

Author: change everywhere.

.. _linux_sources:

Linux installation from sources
===============================

The instructions for installing both the :ref:`Fast DDS library <fastdds_lib_sl>`
and the :ref:`Fast DDS-Gen <fastddsgen_sl>` generation tool from sources are provided in this page.
It is organized as follows:

.. contents::
    :local:
    :backlinks: none
    :depth: 2

.. _fastdds_lib_sl:

Fast DDS library installation
"""""""""""""""""""""""""""""

This section describes the instructions for installing *eProsima Fast DDS* in a Linux environment from
sources. The following packages will be installed:

* :code:`foonathan_memory_vendor`, an STL compatible C++ memory allocator
  `library <https://github.com/foonathan/memory>`_.
* :code:`fastcdr`, a C++ library that serializes according to the
  `standard CDR <https://www.omg.org/cgi-bin/doc?formal/02-06-51>`_ serialization mechanism.
* :code:`fastrtps`, the core library of *eProsima Fast DDS* library.

First of all, the :ref:`requirements_sl` and :ref:`dependencies_sl` detailed below need to be met.

> 

Reviewer: ```suggestion
```

Afterwards, the user can choose whether to follow either the :ref:`colcon <colcon_installation_linux>`)
or the :ref:`CMake <cmake_installation_linux>`) installation instructions.

.. _requirements_sl:

Requirements
------------

The installation of *eProsima Fast DDS* in a Linux environment from binaries requires the following tools to be
installed in the system:

* :ref:`cmake_gcc_pip_wget_git_sl`
* :ref:`gtest_sl` [optional]

.. _cmake_gcc_pip_wget_git_sl:

CMake, g++, pip, wget and git
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

These packages provide the tools required to install *eProsima Fast DDS* and its dependencies from command line.
Install CMake_, `g++ <https://gcc.gnu.org/>`_, pip_, wget_ and git_ using the package manager of the appropriate
Linux distribution. For example, on Ubuntu use the command:

.. code-block:: bash

    sudo apt install cmake g++ pip wget git

.. _gtest_sl:

Gtest
^^^^^

GTest is a unit testing library for C++.

> 

Reviewer: ```suggestion
```

By default, *eProsima Fast DDS* does not compile tests.
It is possible to activate them with the opportune
`CMake configuration options <https://cmake.org/cmake/help/v3.6/manual/cmake.1.html#options>`_
when calling colcon_ or CMake_.
For more details, please refer to the :ref:`cmake_options` section.

> 

Reviewer: There is really no need for all these one-sentence paragraphs
```suggestion
```

For a detailed description of the Gtest installation process, please refer to the
`Gtest Installation Guide <https://github.com/google/googletest>`_.

.. _dependencies_sl:

Dependencies
------------

*eProsima Fast DDS* has the following dependencies, when installed from binaries in a Linux environment:

* :ref:`asiotinyxml2_sl`
* :ref:`openssl_sl`

.. _asiotinyxml2_sl:

Asio and TinyXML2 libraries
^^^^^^^^^^^^^^^^^^^^^^^^^^^

Asio is a cross-platform C++ library for network and low-level I/O programming, which provides a consistent
asynchronous model. TinyXML2 is a simple, small and efficient C++ XML parser.

> 

Reviewer: ```suggestion
```

   * `Fast CDR <https://github.com/eProsima/Fast-CDR.git>`_

     .. code-block:: bash

         cd ~/Fast-DDS
         git clone https://github.com/eProsima/Fast-CDR.git
         mkdir Fast-CDR/build
         cd Fast-CDR/build
         cmake .. -DCMAKE_INSTALL_PREFIX=~/Fast-DDS/install
         sudo cmake --build . --target install

#. Once all dependencies are installed, install *eProsima Fast DDS*:

   .. code-block:: bash

       cd ~/Fast-DDS
       git clone https://github.com/eProsima/Fast-DDS.git
       mkdir Fast-DDS/build
       cd Fast-DDS/build
       cmake ..  -DCMAKE_INSTALL_PREFIX=~/Fast-DDS/install -DCMAKE_PREFIX_PATH=~/Fast-DDS/install
       sudo cmake --build . --target install

.. note::

    By default, *eProsima Fast DDS* does not compile tests.
    However, they can be activated by downloading and installing `Gtest <https://github.com/google/googletest>`_.

.. _global_installation_sl:

Global installation
^^^^^^^^^^^^^^^^^^^

To install *eProsima Fast DDS* system-wide instead of locally, remove all the flags that
appear in the configuration steps of :code:`Fast-CDR` and :code:`Fast-DDS`, and change the first in the

> configuration step of :code:`foonathan_memory_vendor` to the following:

Reviewer: Do not remove `BUILD_SHARED_LIBS` from foonathan

      export LD_LIBRARY_PATH=/usr/local/lib/

* Add it permanently it to the :code:`PATH`, by typing:

  .. code-block:: bash

      echo 'export LD_LIBRARY_PATH=/usr/local/lib/' >> ~/.bashrc

.. _fastddsgen_sl:

Fast DDS-Gen installation
"""""""""""""""""""""""""

This section provides the instructions for installing *Fast DDS-Gen* in a Linux environment from
sources.
*Fast DDS-Gen* is a Java application that generates source code using the data types defined in an IDL file.
Please refer to :ref:`fastrtpsgen_intro` for more information.

Requirements
------------

In order to compile *Fast DDS-Gen*, the following packages need to be installed in the system:

* :ref:`java_sl`
* :ref:`gradle_sl`

.. _java_sl:

Java JDK
^^^^^^^^

The JDK is a development environment for building applications and components using the Java language.
Download and install it at the following the steps given in the
`Oracle website <https://www.oracle.com/java/technologies/javase-downloads.html>`_.

.. _gradle_sl:

Gradle
^^^^^^

Gradle is an open-source build automation tool.
Download and install the last stable version of `Gradle <https://gradle.org/install>`_ in the preferred way.

> Install

Reviewer: This is more compile than install
